crates/wasm-smith/src/core.rs

```diff
@@ -1980,13 +1981,65 @@ impl Module {
+                // For tables, add a new table.
+                wasmparser::types::EntityType::Table(table_type) => {
+                    self.add_arbitrary_table_of_type(table_type.try_into().unwrap(), u)?
+                }
+                // For tags, add the type.
+                wasmparser::types::EntityType::Tag(id) => {
+                    let subtype = exports_types.get(id).unwrap_or_else(|| {
+                        panic!("Unable to get subtype for tag {id:?} in `exports` Wasm")
+                    });
+                    match &subtype.composite_type.inner {
+                        wasmparser::CompositeInnerType::Func(func_type) => {
+                            assert!(
+                                subtype.is_final,
+                                "Subtype {subtype:?} from `exports` Wasm is not final"
+                            );
+                            assert!(
+                                subtype.supertype_idx.is_none(),
+                                "Subtype {subtype:?} from `exports` Wasm has non-empty supertype"
+                            );
+                            assert!(
+                                func_type.results().is_empty(),
+                                "Subtype {subtype:?} for tag {id:?} from `exports` Wasm has non-empty results"
+                            );
+                            let new_type = Rc::new(FuncType {
+                                params: func_type
+                                    .params()
+                                    .iter()
+                                    .copied()
+                                    .map(|t| t.try_into().unwrap())
+                                    .collect(),
+                                results: vec![],
+                            });
+                            self.rec_groups.push(self.types.len()..self.types.len() + 1);
+                            let type_index = self.add_type(SubType {
+                                is_final: true,
+                                supertype: None,
+                                depth: 1,
+                                composite_type: CompositeType::new_func(
+                                    Rc::clone(&new_type),
+                                    subtype.composite_type.shared,
+                                ),
+                            });
+                            let tag_index = self.tags.len() as u32;
+                            self.tags.push(TagType {
+                                func_type_idx: type_index,
+                                func_type: new_type,
+                            });
+                            self.num_defined_tags += 1;
+                            tag_index
+                        }
+                        _ => panic!(
+                            "Unable to handle type {:?} from `exports` Wasm",
+                            subtype.composite_type
+                        ),
+                    }
```

Comment:
This looks like a large copy/paste of what functions are doing above, so could this be refactored to not duplicate the logic?